// @azure/cosmos/dist/commonjs/queryMetrics/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var queryMetrics_exports = {};
__export(queryMetrics_exports, {
  ClientSideMetrics: () => import_clientSideMetrics.ClientSideMetrics,
  QueryMetrics: () => import_queryMetrics.QueryMetrics,
  QueryMetricsConstants: () => import_queryMetricsConstants.default,
  QueryPreparationTimes: () => import_queryPreparationTime.QueryPreparationTimes,
  RuntimeExecutionTimes: () => import_runtimeExecutionTimes.RuntimeExecutionTimes,
  TimeSpan: () => import_timeSpan.TimeSpan
});
module.exports = __toCommonJS(queryMetrics_exports);
var import_clientSideMetrics = require("./clientSideMetrics.js");
var import_queryMetrics = require("./queryMetrics.js");
var import_queryMetricsConstants = __toESM(require("./queryMetricsConstants.js"));
var import_queryPreparationTime = require("./queryPreparationTime.js");
var import_runtimeExecutionTimes = require("./runtimeExecutionTimes.js");
var import_timeSpan = require("./timeSpan.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClientSideMetrics,
  QueryMetrics,
  QueryMetricsConstants,
  QueryPreparationTimes,
  RuntimeExecutionTimes,
  TimeSpan
});
