// @azure/cosmos/dist/commonjs/request/ResourceResponse.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ResourceResponse_exports = {};
__export(ResourceResponse_exports, {
  ResourceResponse: () => ResourceResponse
});
module.exports = __toCommonJS(ResourceResponse_exports);
var import_common = require("../common/index.js");
class ResourceResponse {
  constructor(resource, headers, statusCode, diagnostics, substatus) {
    this.resource = resource;
    this.headers = headers;
    this.statusCode = statusCode;
    this.diagnostics = diagnostics;
    this.substatus = substatus;
  }
  get requestCharge() {
    return Number(this.headers[import_common.Constants.HttpHeaders.RequestCharge]) || 0;
  }
  get activityId() {
    return this.headers[import_common.Constants.HttpHeaders.ActivityId];
  }
  get etag() {
    return this.headers[import_common.Constants.HttpHeaders.ETag];
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ResourceResponse
});
