// @azure/cosmos/dist/commonjs/bulk/ItemOperationContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ItemOperationContext_exports = {};
__export(ItemOperationContext_exports, {
  ItemOperationContext: () => ItemOperationContext
});
module.exports = __toCommonJS(ItemOperationContext_exports);
var import_batch = require("../utils/batch.js");
class ItemOperationContext {
  pkRangeId;
  retryPolicy;
  diagnosticNode;
  taskCompletionSource;
  constructor(pkRangeId, retryPolicy, diagnosticNode) {
    this.pkRangeId = pkRangeId;
    this.retryPolicy = retryPolicy;
    this.diagnosticNode = diagnosticNode;
    this.taskCompletionSource = new import_batch.TaskCompletionSource();
  }
  get operationPromise() {
    return this.taskCompletionSource.task;
  }
  addDiagnosticChild(diagnosticNode) {
    this.diagnosticNode.addBulkChildNode(diagnosticNode, this.diagnosticNode.diagnosticLevel);
    diagnosticNode.updateTimestamp();
  }
  updatePKRangeId(pkRangeId) {
    this.pkRangeId = pkRangeId;
  }
  complete(result) {
    this.taskCompletionSource.setResult(result);
  }
  fail(error) {
    this.taskCompletionSource.setException(error.error);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ItemOperationContext
});
