// @azure/cosmos/dist/commonjs/client/Script/Scripts.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Scripts_exports = {};
__export(Scripts_exports, {
  Scripts: () => Scripts
});
module.exports = __toCommonJS(Scripts_exports);
var import_StoredProcedure = require("../StoredProcedure/index.js");
var import_Trigger = require("../Trigger/index.js");
var import_UserDefinedFunction = require("../UserDefinedFunction/index.js");
class Scripts {
  /**
   * @param container - The parent {@link Container}.
   * @hidden
   */
  constructor(container, clientContext) {
    this.container = container;
    this.clientContext = clientContext;
  }
  /**
   * Used to read, replace, or delete a specific, existing {@link StoredProcedure} by id.
   *
   * Use `.storedProcedures` for creating new stored procedures, or querying/reading all stored procedures.
   * @param id - The id of the {@link StoredProcedure}.
   */
  storedProcedure(id) {
    return new import_StoredProcedure.StoredProcedure(this.container, id, this.clientContext);
  }
  /**
   * Used to read, replace, or delete a specific, existing {@link Trigger} by id.
   *
   * Use `.triggers` for creating new triggers, or querying/reading all triggers.
   * @param id - The id of the {@link Trigger}.
   */
  trigger(id) {
    return new import_Trigger.Trigger(this.container, id, this.clientContext);
  }
  /**
   * Used to read, replace, or delete a specific, existing {@link UserDefinedFunction} by id.
   *
   * Use `.userDefinedFunctions` for creating new user defined functions, or querying/reading all user defined functions.
   * @param id - The id of the {@link UserDefinedFunction}.
   */
  userDefinedFunction(id) {
    return new import_UserDefinedFunction.UserDefinedFunction(this.container, id, this.clientContext);
  }
  $sprocs;
  /**
   * Operations for creating new stored procedures, and reading/querying all stored procedures.
   *
   * For reading, replacing, or deleting an existing stored procedure, use `.storedProcedure(id)`.
   */
  get storedProcedures() {
    if (!this.$sprocs) {
      this.$sprocs = new import_StoredProcedure.StoredProcedures(this.container, this.clientContext);
    }
    return this.$sprocs;
  }
  $triggers;
  /**
   * Operations for creating new triggers, and reading/querying all triggers.
   *
   * For reading, replacing, or deleting an existing trigger, use `.trigger(id)`.
   */
  get triggers() {
    if (!this.$triggers) {
      this.$triggers = new import_Trigger.Triggers(this.container, this.clientContext);
    }
    return this.$triggers;
  }
  $udfs;
  /**
   * Operations for creating new user defined functions, and reading/querying all user defined functions.
   *
   * For reading, replacing, or deleting an existing user defined function, use `.userDefinedFunction(id)`.
   */
  get userDefinedFunctions() {
    if (!this.$udfs) {
      this.$udfs = new import_UserDefinedFunction.UserDefinedFunctions(this.container, this.clientContext);
    }
    return this.$udfs;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Scripts
});
