// @azure/cosmos/dist/commonjs/documents/ConnectionPolicy.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ConnectionPolicy_exports = {};
__export(ConnectionPolicy_exports, {
  defaultConnectionPolicy: () => defaultConnectionPolicy
});
module.exports = __toCommonJS(ConnectionPolicy_exports);
var import_constants = require("../common/constants.js");
var import_ConnectionMode = require("./ConnectionMode.js");
const defaultConnectionPolicy = Object.freeze({
  connectionMode: import_ConnectionMode.ConnectionMode.Gateway,
  requestTimeout: 6e4,
  enableEndpointDiscovery: true,
  preferredLocations: [],
  retryOptions: {
    maxRetryAttemptCount: import_constants.Constants.ThrottledRequestMaxRetryAttemptCount,
    fixedRetryIntervalInMilliseconds: import_constants.Constants.ThrottledRequestFixedRetryIntervalInMs,
    maxWaitTimeInSeconds: import_constants.Constants.ThrottledRequestMaxWaitTimeInSeconds
  },
  useMultipleWriteLocations: true,
  endpointRefreshRateInMs: 3e5,
  enableBackgroundEndpointRefreshing: true,
  enablePartitionLevelFailover: true,
  enablePartitionLevelCircuitBreaker: false
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  defaultConnectionPolicy
});
