// @azure/cosmos/dist/commonjs/documents/ConsistencyLevel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ConsistencyLevel_exports = {};
__export(ConsistencyLevel_exports, {
  ConsistencyLevel: () => ConsistencyLevel
});
module.exports = __toCommonJS(ConsistencyLevel_exports);
var ConsistencyLevel = /* @__PURE__ */ ((ConsistencyLevel2) => {
  ConsistencyLevel2["Strong"] = "Strong";
  ConsistencyLevel2["BoundedStaleness"] = "BoundedStaleness";
  ConsistencyLevel2["Session"] = "Session";
  ConsistencyLevel2["Eventual"] = "Eventual";
  ConsistencyLevel2["ConsistentPrefix"] = "ConsistentPrefix";
  return ConsistencyLevel2;
})(ConsistencyLevel || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ConsistencyLevel
});
