// @azure/cosmos/dist/commonjs/documents/IndexingPolicy.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var IndexingPolicy_exports = {};
__export(IndexingPolicy_exports, {
  SpatialType: () => SpatialType,
  VectorIndexType: () => VectorIndexType
});
module.exports = __toCommonJS(IndexingPolicy_exports);
var SpatialType = /* @__PURE__ */ ((SpatialType2) => {
  SpatialType2["LineString"] = "LineString";
  SpatialType2["MultiPolygon"] = "MultiPolygon";
  SpatialType2["Point"] = "Point";
  SpatialType2["Polygon"] = "Polygon";
  return SpatialType2;
})(SpatialType || {});
var VectorIndexType = /* @__PURE__ */ ((VectorIndexType2) => {
  VectorIndexType2["Flat"] = "flat";
  VectorIndexType2["DiskANN"] = "diskANN";
  VectorIndexType2["QuantizedFlat"] = "quantizedFlat";
  return VectorIndexType2;
})(VectorIndexType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SpatialType,
  VectorIndexType
});
