// @azure/cosmos/dist/commonjs/documents/TriggerOperation.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TriggerOperation_exports = {};
__export(TriggerOperation_exports, {
  TriggerOperation: () => TriggerOperation
});
module.exports = __toCommonJS(TriggerOperation_exports);
var TriggerOperation = /* @__PURE__ */ ((TriggerOperation2) => {
  TriggerOperation2["All"] = "all";
  TriggerOperation2["Create"] = "create";
  TriggerOperation2["Update"] = "update";
  TriggerOperation2["Delete"] = "delete";
  TriggerOperation2["Replace"] = "replace";
  return TriggerOperation2;
})(TriggerOperation || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TriggerOperation
});
