// @azure/cosmos/dist/commonjs/documents/VectorEmbeddingPolicy.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VectorEmbeddingPolicy_exports = {};
__export(VectorEmbeddingPolicy_exports, {
  VectorEmbeddingDataType: () => VectorEmbeddingDataType,
  VectorEmbeddingDistanceFunction: () => VectorEmbeddingDistanceFunction
});
module.exports = __toCommonJS(VectorEmbeddingPolicy_exports);
var VectorEmbeddingDataType = /* @__PURE__ */ ((VectorEmbeddingDataType2) => {
  VectorEmbeddingDataType2["Float16"] = "float16";
  VectorEmbeddingDataType2["Float32"] = "float32";
  VectorEmbeddingDataType2["UInt8"] = "uint8";
  VectorEmbeddingDataType2["Int8"] = "int8";
  return VectorEmbeddingDataType2;
})(VectorEmbeddingDataType || {});
var VectorEmbeddingDistanceFunction = /* @__PURE__ */ ((VectorEmbeddingDistanceFunction2) => {
  VectorEmbeddingDistanceFunction2["Euclidean"] = "euclidean";
  VectorEmbeddingDistanceFunction2["Cosine"] = "cosine";
  VectorEmbeddingDistanceFunction2["DotProduct"] = "dotproduct";
  return VectorEmbeddingDistanceFunction2;
})(VectorEmbeddingDistanceFunction || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VectorEmbeddingDataType,
  VectorEmbeddingDistanceFunction
});
