// @azure/cosmos/dist/commonjs/queryExecutionContext/Aggregators/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Aggregators_exports = {};
__export(Aggregators_exports, {
  AverageAggregator: () => import_AverageAggregator.AverageAggregator,
  CountAggregator: () => import_CountAggregator.CountAggregator,
  MaxAggregator: () => import_MaxAggregator.MaxAggregator,
  MinAggregator: () => import_MinAggregator.MinAggregator,
  SumAggregator: () => import_SumAggregator.SumAggregator,
  createAggregator: () => createAggregator
});
module.exports = __toCommonJS(Aggregators_exports);
var import_AverageAggregator = require("./AverageAggregator.js");
var import_CountAggregator = require("./CountAggregator.js");
var import_MaxAggregator = require("./MaxAggregator.js");
var import_MinAggregator = require("./MinAggregator.js");
var import_SumAggregator = require("./SumAggregator.js");
var import_StaticValueAggregator = require("./StaticValueAggregator.js");
var import_MakeListAggregator = require("./MakeListAggregator.js");
var import_MakeSetAggregator = require("./MakeSetAggregator.js");
function createAggregator(aggregateType) {
  switch (aggregateType) {
    case "Average":
      return new import_AverageAggregator.AverageAggregator();
    case "Count":
    case "CountIf":
      return new import_CountAggregator.CountAggregator();
    case "Max":
      return new import_MaxAggregator.MaxAggregator();
    case "Min":
      return new import_MinAggregator.MinAggregator();
    case "Sum":
      return new import_SumAggregator.SumAggregator();
    case "MakeList":
      return new import_MakeListAggregator.MakeListAggregator();
    case "MakeSet":
      return new import_MakeSetAggregator.MakeSetAggregator();
    default:
      return new import_StaticValueAggregator.StaticValueAggregator();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AverageAggregator,
  CountAggregator,
  MaxAggregator,
  MinAggregator,
  SumAggregator,
  createAggregator
});
