// @azure/cosmos/dist/commonjs/client/ChangeFeed/ChangeFeedRange.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChangeFeedRange_exports = {};
__export(ChangeFeedRange_exports, {
  ChangeFeedRange: () => ChangeFeedRange
});
module.exports = __toCommonJS(ChangeFeedRange_exports);
class ChangeFeedRange {
  /**
   * Min value for the range.
   */
  minInclusive;
  /**
   * Max value for the range.
   */
  maxExclusive;
  /**
   * Continuation token from where to start reading changes.
   */
  continuationToken;
  /**
   * Min epk value to begin reading changes from in case changefeed of entire partition is not requested.
   */
  epkMinHeader;
  /**
   * Max epk value to begin reading changes from in case changefeed of entire partition is not requested.
   */
  epkMaxHeader;
  constructor(minInclusive, maxExclusive, continuationToken, epkMinHeader, epkMaxHeader) {
    this.minInclusive = minInclusive;
    this.maxExclusive = maxExclusive;
    this.continuationToken = continuationToken;
    this.epkMinHeader = epkMinHeader;
    this.epkMaxHeader = epkMaxHeader;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ChangeFeedRange
});
