// @azure/cosmos/dist/commonjs/client/ChangeFeed/ChangeFeedIteratorBuilder.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChangeFeedIteratorBuilder_exports = {};
__export(ChangeFeedIteratorBuilder_exports, {
  ChangeFeedIteratorBuilder: () => ChangeFeedIteratorBuilder
});
module.exports = __toCommonJS(ChangeFeedIteratorBuilder_exports);
var import_buildChangeFeedIterator = require("./buildChangeFeedIterator.js");
var import_request = require("../../request/index.js");
class ChangeFeedIteratorBuilder {
  /**
   * @internal
   */
  constructor(cfOptions, clientContext, container, partitionKeyRangeCache) {
    this.cfOptions = cfOptions;
    this.clientContext = clientContext;
    this.container = container;
    this.partitionKeyRangeCache = partitionKeyRangeCache;
    this.isInitialized = false;
  }
  iterator;
  isInitialized;
  /**
   * Change feed is an infinite feed. hasMoreResults is always true.
   */
  get hasMoreResults() {
    return true;
  }
  /**
   * Gets an async iterator which will yield change feed results.
   */
  async *getAsyncIterator() {
    await this.initializeIterator();
    do {
      const result = await this.iterator.readNext();
      yield result;
    } while (this.hasMoreResults);
  }
  /**
   * Returns the result of change feed from Azure Cosmos DB.
   */
  async readNext() {
    await this.initializeIterator();
    return this.iterator.readNext();
  }
  async initializeIterator() {
    if (!this.isInitialized) {
      try {
        const iterator = await (0, import_buildChangeFeedIterator.buildChangeFeedIterator)(
          this.cfOptions,
          this.clientContext,
          this.container,
          this.partitionKeyRangeCache
        );
        this.isInitialized = true;
        this.iterator = iterator;
      } catch (err) {
        throw new import_request.ErrorResponse(err.message);
      }
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ChangeFeedIteratorBuilder
});
