// @azure/cosmos/dist/commonjs/client/ChangeFeed/FeedRangeQueue.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FeedRangeQueue_exports = {};
__export(FeedRangeQueue_exports, {
  FeedRangeQueue: () => FeedRangeQueue
});
module.exports = __toCommonJS(FeedRangeQueue_exports);
var import_ChangeFeedRange = require("./ChangeFeedRange.js");
class FeedRangeQueue {
  elements;
  constructor() {
    this.elements = [];
  }
  modifyFirstElement(newItem) {
    if (!this.isEmpty()) {
      this.elements[0] = newItem;
    }
  }
  enqueue(item) {
    this.elements.push(item);
  }
  dequeue() {
    return this.elements.shift();
  }
  peek() {
    return !this.isEmpty() ? this.elements[0] : void 0;
  }
  isEmpty() {
    return this.elements.length === 0;
  }
  moveFirstElementToTheEnd() {
    if (!this.isEmpty()) {
      this.elements.push(this.dequeue());
    }
  }
  /**
   * Returns a snapshot of the queue as an array to be used as Continuation token.
   */
  returnSnapshot() {
    const allFeedRanges = [];
    this.elements.map((element) => {
      const minInclusive = element.epkMinHeader ? element.epkMinHeader : element.minInclusive;
      const maxExclusive = element.epkMaxHeader ? element.epkMaxHeader : element.maxExclusive;
      const feedRangeElement = new import_ChangeFeedRange.ChangeFeedRange(
        minInclusive,
        maxExclusive,
        element.continuationToken
      );
      allFeedRanges.push(feedRangeElement);
    });
    return allFeedRanges;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FeedRangeQueue
});
