// @azure/cosmos/dist/commonjs/client/ChangeFeed/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChangeFeed_exports = {};
__export(ChangeFeed_exports, {
  ChangeFeedForEpkRange: () => import_ChangeFeedForEpkRange.ChangeFeedForEpkRange,
  ChangeFeedForPartitionKey: () => import_ChangeFeedForPartitionKey.ChangeFeedForPartitionKey,
  ChangeFeedIteratorBuilder: () => import_ChangeFeedIteratorBuilder.ChangeFeedIteratorBuilder,
  ChangeFeedIteratorResponse: () => import_ChangeFeedIteratorResponse.ChangeFeedIteratorResponse,
  ChangeFeedMode: () => import_ChangeFeedMode.ChangeFeedMode,
  ChangeFeedPolicy: () => import_ChangeFeedPolicy.ChangeFeedPolicy,
  ChangeFeedRetentionTimeSpan: () => import_ChangeFeedRetentionTimeSpan.ChangeFeedRetentionTimeSpan,
  ChangeFeedStartFrom: () => import_ChangeFeedStartFrom.ChangeFeedStartFrom,
  FeedRange: () => import_FeedRange.FeedRange,
  FeedRangeInternal: () => import_FeedRange.FeedRangeInternal,
  buildChangeFeedIterator: () => import_buildChangeFeedIterator.buildChangeFeedIterator,
  buildInternalChangeFeedOptions: () => import_changeFeedUtils.buildInternalChangeFeedOptions
});
module.exports = __toCommonJS(ChangeFeed_exports);
var import_ChangeFeedForEpkRange = require("./ChangeFeedForEpkRange.js");
var import_ChangeFeedForPartitionKey = require("./ChangeFeedForPartitionKey.js");
var import_ChangeFeedIteratorResponse = require("./ChangeFeedIteratorResponse.js");
var import_FeedRange = require("./FeedRange.js");
var import_ChangeFeedStartFrom = require("./ChangeFeedStartFrom.js");
var import_ChangeFeedMode = require("./ChangeFeedMode.js");
var import_changeFeedUtils = require("./changeFeedUtils.js");
var import_buildChangeFeedIterator = require("./buildChangeFeedIterator.js");
var import_ChangeFeedRetentionTimeSpan = require("./ChangeFeedRetentionTimeSpan.js");
var import_ChangeFeedPolicy = require("./ChangeFeedPolicy.js");
var import_ChangeFeedIteratorBuilder = require("./ChangeFeedIteratorBuilder.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ChangeFeedForEpkRange,
  ChangeFeedForPartitionKey,
  ChangeFeedIteratorBuilder,
  ChangeFeedIteratorResponse,
  ChangeFeedMode,
  ChangeFeedPolicy,
  ChangeFeedRetentionTimeSpan,
  ChangeFeedStartFrom,
  FeedRange,
  FeedRangeInternal,
  buildChangeFeedIterator,
  buildInternalChangeFeedOptions
});
