// @azure/cosmos/dist/commonjs/client/UserDefinedFunction/UserDefinedFunctions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var UserDefinedFunctions_exports = {};
__export(UserDefinedFunctions_exports, {
  UserDefinedFunctions: () => UserDefinedFunctions
});
module.exports = __toCommonJS(UserDefinedFunctions_exports);
var import_common = require("../../common/index.js");
var import_queryIterator = require("../../queryIterator.js");
var import_UserDefinedFunction = require("./UserDefinedFunction.js");
var import_UserDefinedFunctionResponse = require("./UserDefinedFunctionResponse.js");
var import_diagnostics = require("../../utils/diagnostics.js");
class UserDefinedFunctions {
  /**
   * @hidden
   * @param container - The parent {@link Container}.
   */
  constructor(container, clientContext) {
    this.container = container;
    this.clientContext = clientContext;
  }
  query(query, options) {
    const path = (0, import_common.getPathFromLink)(this.container.url, import_common.ResourceType.udf);
    const id = (0, import_common.getIdFromLink)(this.container.url);
    return new import_queryIterator.QueryIterator(this.clientContext, query, options, (diagnosticNode, innerOptions) => {
      return this.clientContext.queryFeed({
        path,
        resourceType: import_common.ResourceType.udf,
        resourceId: id,
        resultFn: (result) => result.UserDefinedFunctions,
        query,
        options: innerOptions,
        diagnosticNode
      });
    });
  }
  /**
   * Read all User Defined Functions.
   * @example Read all User Defined Functions to array.
   * ```ts snippet:UserDefinedFunctionsReadAll
   * import { CosmosClient } from "@azure/cosmos";
   *
   * const endpoint = "https://your-account.documents.azure.com";
   * const key = "<database account masterkey>";
   * const client = new CosmosClient({ endpoint, key });
   *
   * const { database } = await client.databases.createIfNotExists({ id: "Test Database" });
   *
   * const { container } = await database.containers.createIfNotExists({ id: "Test Container" });
   *
   * const { resources: udfList } = await container.scripts.userDefinedFunctions.readAll().fetchAll();
   * ```
   */
  readAll(options) {
    return this.query(void 0, options);
  }
  /**
   * Create a UserDefinedFunction.
   *
   * Azure Cosmos DB supports JavaScript UDFs which can be used inside queries, stored procedures and triggers.
   *
   * For additional details, refer to the server-side JavaScript API documentation.
   * @example
   * ```ts snippet:UserDefinedFunctionsCreate
   * import { CosmosClient, UserDefinedFunctionDefinition } from "@azure/cosmos";
   *
   * const endpoint = "https://your-account.documents.azure.com";
   * const key = "<database account masterkey>";
   * const client = new CosmosClient({ endpoint, key });
   * const { database } = await client.databases.createIfNotExists({ id: "Test Database" });
   * const { container } = await database.containers.createIfNotExists({ id: "Test Container" });
   *
   * const udfDefinition: UserDefinedFunctionDefinition = {
   *   id: "sample udf",
   *   body: "function () { const x = 10; }",
   * };
   *
   * const { resource: udf } = await container.scripts.userDefinedFunctions.create(udfDefinition);
   * ```
   */
  async create(body, options) {
    return (0, import_diagnostics.withDiagnostics)(async (diagnosticNode) => {
      if (body.body) {
        body.body = body.body.toString();
      }
      const err = {};
      if (!(0, import_common.isResourceValid)(body, err)) {
        throw err;
      }
      const path = (0, import_common.getPathFromLink)(this.container.url, import_common.ResourceType.udf);
      const id = (0, import_common.getIdFromLink)(this.container.url);
      const response = await this.clientContext.create({
        body,
        path,
        resourceType: import_common.ResourceType.udf,
        resourceId: id,
        options,
        diagnosticNode
      });
      const ref = new import_UserDefinedFunction.UserDefinedFunction(this.container, response.result.id, this.clientContext);
      return new import_UserDefinedFunctionResponse.UserDefinedFunctionResponse(
        response.result,
        response.headers,
        response.code,
        ref,
        (0, import_diagnostics.getEmptyCosmosDiagnostics)()
      );
    }, this.clientContext);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  UserDefinedFunctions
});
