// @azure/cosmos/dist/commonjs/client/Trigger/Trigger.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Trigger_exports = {};
__export(Trigger_exports, {
  Trigger: () => Trigger
});
module.exports = __toCommonJS(Trigger_exports);
var import_common = require("../../common/index.js");
var import_TriggerResponse = require("./TriggerResponse.js");
var import_diagnostics = require("../../utils/diagnostics.js");
class Trigger {
  /**
   * @hidden
   * @param container - The parent {@link Container}.
   * @param id - The id of the given {@link Trigger}.
   */
  constructor(container, id, clientContext) {
    this.container = container;
    this.id = id;
    this.clientContext = clientContext;
  }
  /**
   * Returns a reference URL to the resource. Used for linking in Permissions.
   */
  get url() {
    return (0, import_common.createTriggerUri)(this.container.database.id, this.container.id, this.id);
  }
  /**
   * Read the {@link TriggerDefinition} for the given {@link Trigger}.
   * @example
   * ```ts snippet:TriggerRead
   * import { CosmosClient } from "@azure/cosmos";
   *
   * const endpoint = "https://your-account.documents.azure.com";
   * const key = "<database account masterkey>";
   * const client = new CosmosClient({ endpoint, key });
   * const { database } = await client.databases.createIfNotExists({ id: "Test Database" });
   * const { container } = await database.containers.createIfNotExists({ id: "Test Container" });
   *
   * const { resource: trigger } = await container.scripts.trigger("<trigger-id>").read();
   * ```
   */
  async read(options) {
    return (0, import_diagnostics.withDiagnostics)(async (diagnosticNode) => {
      const path = (0, import_common.getPathFromLink)(this.url);
      const id = (0, import_common.getIdFromLink)(this.url);
      const response = await this.clientContext.read({
        path,
        resourceType: import_common.ResourceType.trigger,
        resourceId: id,
        options,
        diagnosticNode
      });
      return new import_TriggerResponse.TriggerResponse(
        response.result,
        response.headers,
        response.code,
        this,
        (0, import_diagnostics.getEmptyCosmosDiagnostics)()
      );
    }, this.clientContext);
  }
  /**
   * Replace the given {@link Trigger} with the specified {@link TriggerDefinition}.
   * @param body - The specified {@link TriggerDefinition} to replace the existing definition with.
   * @example
   * ```ts snippet:TriggerReplace
   * import { CosmosClient, TriggerDefinition, TriggerType, TriggerOperation } from "@azure/cosmos";
   *
   * const endpoint = "https://your-account.documents.azure.com";
   * const key = "<database account masterkey>";
   * const client = new CosmosClient({ endpoint, key });
   * const { database } = await client.databases.createIfNotExists({ id: "Test Database" });
   * const { container } = await database.containers.createIfNotExists({ id: "Test Container" });
   *
   * const triggerDefinition: TriggerDefinition = {
   *   id: "sample trigger",
   *   body: "serverScript() { var x = 10; }",
   *   triggerType: TriggerType.Pre,
   *   triggerOperation: TriggerOperation.All,
   * };
   *
   * const { resource: trigger } = await container.scripts.triggers.create(triggerDefinition);
   *
   * trigger.body = "function () { const x = 20; console.log(x); }";
   * const { resource: replacedTrigger } = await container.scripts.trigger(trigger.id).replace(trigger);
   * ```
   */
  async replace(body, options) {
    return (0, import_diagnostics.withDiagnostics)(async (diagnosticNode) => {
      if (body.body) {
        body.body = body.body.toString();
      }
      const err = {};
      if (!(0, import_common.isResourceValid)(body, err)) {
        throw err;
      }
      const path = (0, import_common.getPathFromLink)(this.url);
      const id = (0, import_common.getIdFromLink)(this.url);
      const response = await this.clientContext.replace({
        body,
        path,
        resourceType: import_common.ResourceType.trigger,
        resourceId: id,
        options,
        diagnosticNode
      });
      return new import_TriggerResponse.TriggerResponse(
        response.result,
        response.headers,
        response.code,
        this,
        (0, import_diagnostics.getEmptyCosmosDiagnostics)()
      );
    }, this.clientContext);
  }
  /**
   * Delete the given {@link Trigger}.
   * @example
   * ```ts snippet:TriggerDelete
   * import { CosmosClient } from "@azure/cosmos";
   *
   * const endpoint = "https://your-account.documents.azure.com";
   * const key = "<database account masterkey>";
   * const client = new CosmosClient({ endpoint, key });
   * const { database } = await client.databases.createIfNotExists({ id: "Test Database" });
   * const { container } = await database.containers.createIfNotExists({ id: "Test Container" });
   *
   * await container.scripts.trigger("<trigger-id>").delete();
   * ```
   */
  async delete(options) {
    return (0, import_diagnostics.withDiagnostics)(async (diagnosticNode) => {
      const path = (0, import_common.getPathFromLink)(this.url);
      const id = (0, import_common.getIdFromLink)(this.url);
      const response = await this.clientContext.delete({
        path,
        resourceType: import_common.ResourceType.trigger,
        resourceId: id,
        options,
        diagnosticNode
      });
      return new import_TriggerResponse.TriggerResponse(
        response.result,
        response.headers,
        response.code,
        this,
        (0, import_diagnostics.getEmptyCosmosDiagnostics)()
      );
    }, this.clientContext);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Trigger
});
