// @azure/cosmos/dist/commonjs/encryption/enums/TypeMarker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TypeMarker_exports = {};
__export(TypeMarker_exports, {
  TypeMarker: () => TypeMarker
});
module.exports = __toCommonJS(TypeMarker_exports);
var TypeMarker = /* @__PURE__ */ ((TypeMarker2) => {
  TypeMarker2[TypeMarker2["Null"] = 1] = "Null";
  TypeMarker2[TypeMarker2["Boolean"] = 2] = "Boolean";
  TypeMarker2[TypeMarker2["Double"] = 3] = "Double";
  TypeMarker2[TypeMarker2["Long"] = 4] = "Long";
  TypeMarker2[TypeMarker2["String"] = 5] = "String";
  return TypeMarker2;
})(TypeMarker || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TypeMarker
});
