// @azure/cosmos/dist/commonjs/encryption/EncryptionSettings.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EncryptionSettings_exports = {};
__export(EncryptionSettings_exports, {
  EncryptionSettings: () => EncryptionSettings
});
module.exports = __toCommonJS(EncryptionSettings_exports);
class EncryptionSettings {
  id;
  // databaseRid + '/' + containerRid
  containerRid;
  partitionKeyPaths;
  pathsToEncrypt = [];
  // key is property path
  encryptionSettingForProperties = {};
  // getContainerRid
  constructor(id, containerRid, partitionKeyPaths) {
    this.id = id;
    this.containerRid = containerRid;
    this.partitionKeyPaths = partitionKeyPaths;
  }
  setEncryptionSettingForProperty(key, encryptionSettingForProperty) {
    this.encryptionSettingForProperties[key] = encryptionSettingForProperty;
  }
  getEncryptionSettingForProperty(propertyName) {
    return this.encryptionSettingForProperties[propertyName];
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EncryptionSettings
});
