// @azure/cosmos/dist/commonjs/encryption/KeyEncryptionKey.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var KeyEncryptionKey_exports = {};
__export(KeyEncryptionKey_exports, {
  KeyEncryptionKey: () => KeyEncryptionKey
});
module.exports = __toCommonJS(KeyEncryptionKey_exports);
var import_KeyEncryptionAlgorithm = require("./enums/KeyEncryptionAlgorithm.js");
class KeyEncryptionKey {
  encryptionAlgorithm;
  name;
  path;
  keyStoreProvider;
  constructor(name, path, keyStoreProvider) {
    this.name = name;
    this.path = path;
    this.keyStoreProvider = keyStoreProvider;
    this.encryptionAlgorithm = import_KeyEncryptionAlgorithm.KeyEncryptionAlgorithm.RSA_OAEP;
  }
  async wrapEncryptionKey(plainTextEncryptionKey) {
    return this.keyStoreProvider.wrapKey(
      this.path,
      this.encryptionAlgorithm,
      plainTextEncryptionKey
    );
  }
  async unwrapEncryptionKey(wrappedEncryptionKey) {
    return this.keyStoreProvider.unwrapKey(
      this.path,
      this.encryptionAlgorithm,
      wrappedEncryptionKey
    );
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  KeyEncryptionKey
});
